File: frontend/src/data/leave.ts
import { computed, toValue, watch, type MaybeRefOrGetter } from 'vue'
import { useCall } from 'frappe-ui'
import {
  decisionButton as buildDecisionButton,
  decisionButtons as buildDecisionButtons,
  styleTheme,
  type BadgeTheme,
  type Decision,
  type DecisionButton,
} from './workflowStyle'

export type { DecisionButton }

/**
 * The employee behind this login, as the leave pages use it.
 *
 * Two fields, because two is what they read: the record a request is raised
 * against, and the approver it defaults to. The old endpoint fetched seven and
 * derived an eighth, and nothing rendered any of the other six.
 */
export interface MyEmployee {
  name: string
  leave_approver: string | null
}

/** What to ask the document API for. Not a permission boundary — Frappe drops
 *  fields the caller may not read from the select on its own — so the page is
 *  free to name the two it draws. */
const EMPLOYEE_FIELDS = ['name', 'leave_approver']

/** Likewise for an application row: exactly what `MyLeave` puts on screen. */
const APPLICATION_FIELDS = [
  'name',
  'leave_type',
  'description',
  'posting_date',
  'from_date',
  'to_date',
  'total_leave_days',
  'status',
  'docstatus',
]

/**
 * One outcome an approver may be offered. The shared shape -- see `Decision` in
 * `workflowStyle`, which procurement's and profile's vocabularies use too.
 */
export type LeaveDecision = Decision

/**
 * Why there is no employee record, when `myEmployee` is null.
 *
 * `missing` sends the reader to HR and `forbidden` to whoever administers
 * permissions, which is the whole reason the server distinguishes them — one
 * message for both told a user whose access had been revoked to ask HR to
 * create a record that already names them. See `session_employee_access`.
 */
export type EmployeeAccess = 'visible' | 'forbidden' | 'missing'

export interface LeaveWorkflowState {
  state: string
  doc_status: number
  style: string | null
}

export interface LeaveWorkflow {
  name: string
  workflow_state_field: string
  states: LeaveWorkflowState[]
  transitions: { state: string; action: string; next_state: string }[]
}

export interface LeavePermissions {
  read: boolean
  request: boolean
  /** What makes an Employee row this session's own. The server's predicate, not
   *  this page's: `user_id` names the login and `status` rules out a leaver, and
   *  a filter written here would be the second place that rule lived. */
  employee_filters: Record<string, string>
  employee_access: EmployeeAccess
  /** The active Workflow, if the site runs one. Its states and their styling are
   *  what a row fetched through the document API is labelled from. */
  workflow: LeaveWorkflow | null
  approve: boolean
  pending_approvals: number
  /** The outcomes the server will accept, in the order they should be offered. */
  decisions: LeaveDecision[]
  /** How many rows a queue returns. The badge counts to the same ceiling. */
  page_length: number
  /** Whether HR Settings makes the approver mandatory. The form asks rather
   *  than assuming: a site that made it optional is one where a blank field
   *  must be allowed through. */
  approver_mandatory: boolean
  /** The link query that resolves candidate approvers, named by the server so a
   *  site can point it somewhere else. */
  approver_query: string
}

/** One row of a leave application list. */
export interface LeaveApplicationRow {
  name: string
  employee: string
  employee_name: string
  leave_type: string
  from_date: string
  to_date: string
  total_leave_days: number
  half_day: 0 | 1
  description: string | null
  status: string
  docstatus: 0 | 1 | 2
  leave_approver: string | null
  leave_approver_name: string | null
  posting_date: string
  /** How this row reads, in the server's words and the site's styling.
   *
   *  Present only on approvals-queue rows, which the server decorates because
   *  what a reviewer may do to a row is a permission question anyway. Rows the
   *  page fetches for itself arrive without them and are labelled by
   *  `leaveStatus`, from the same workflow states and decision styles the
   *  server would have used. */
  status_label?: string
  status_style?: string | null
  /** Whether *this* user may decide *this* application. Only the approvals
   *  queue fills it in; the server settles it again before writing anything. */
  can_decide?: boolean
  /** The outcomes this user may apply to *this* row, named as the server names
   *  them. Transitions from the active Workflow where a site runs one. */
  actions?: string[]
  /** The workflow's state column, whatever a site named it. Not knowable
   *  statically, which is why it is reached by index rather than by name. */
  [key: string]: unknown
}

const leavePermissionsCall = useCall<LeavePermissions>({
  url: '/api/v2/method/tbs_commons.leave.api.get_leave_permissions',
})

const NO_LEAVE_PERMISSIONS: LeavePermissions = {
  read: false,
  request: false,
  employee_filters: {},
  // Before the answer is in, say nothing rather than accuse anyone: `missing`
  // is the state the page renders with no explanation attached to it.
  employee_access: 'missing',
  workflow: null,
  approve: false,
  pending_approvals: 0,
  decisions: [],
  page_length: 0,
  approver_mandatory: true,
  approver_query: '',
}

export const leaveCan = computed(
  () => leavePermissionsCall.data ?? NO_LEAVE_PERMISSIONS
)

/**
 * Whether the answer is in — settled or refused, not merely arrived.
 *
 * Deliberately not `data != null`: a call that fails never sets `data`, and a
 * page gating its skeleton on that sits in the skeleton for ever with nothing
 * to show for it. `leavePermissionsError` is what it should say instead.
 */
export const leavePermissionsLoaded = computed(
  () => leavePermissionsCall.isFinished
)

export const leavePermissionsError = computed(
  () => leavePermissionsCall.error ?? null
)

export function reloadLeavePermissions() {
  return leavePermissionsCall.reload()
}

/**
 * The Employee record behind the session user, through Frappe's document API.
 *
 * Every permission the site has configured applies on the way in — role
 * permissions, User Permissions, and this app's own gate — without this module
 * restating any of them. The filter is the server's (`employee_filters`), so
 * what counts as "mine" is still settled in one place.
 *
 * Nothing is asked for unless the server has already said there is a readable
 * record. `employee_access` is that answer, and firing the read regardless
 * would spend a round trip to rediscover it.
 */
const myEmployeeCall = useCall<
  MyEmployee[],
  { filters: string; fields: string; limit: string }
>({
  url: '/api/v2/document/Employee',
  params: () => ({
    filters: JSON.stringify(leaveCan.value.employee_filters),
    fields: JSON.stringify(EMPLOYEE_FIELDS),
    limit: '1',
  }),
  immediate: false,
})

watch(
  () => leaveCan.value.employee_access,
  (access) => {
    if (access === 'visible') myEmployeeCall.reload()
  },
  { immediate: true }
)

/** The Employee record behind the session user — `null` when none is readable. */
export const myEmployee = computed(() => myEmployeeCall.data?.[0] ?? null)

/**
 * Both answers in.
 *
 * There is nothing to wait for when the server has already said the record is
 * absent or withheld: no read was issued, so `isFinished` would never come.
 */
export const myEmployeeLoaded = computed(
  () =>
    leavePermissionsLoaded.value &&
    (leaveCan.value.employee_access !== 'visible' || myEmployeeCall.isFinished)
)

/**
 * The session user's own leave applications, newest first.
 *
 * Filtered on the employee rather than the login, because leave is requested
 * against an employee record. The read waits for that employee: an unfiltered
 * one would be everybody's, which is exactly what an approver would get.
 */
export function useMyLeaveApplications() {
  const applications = useCall<
    LeaveApplicationRow[],
    { filters: string; fields: string; order_by: string; limit: string }
  >({
    url: '/api/v2/document/Leave Application',
    params: () => ({
      filters: JSON.stringify({ employee: myEmployee.value?.name }),
      // The workflow's state column where a site runs one, so a row can be
      // labelled with the state rather than with `status`.
      fields: JSON.stringify(
        leaveCan.value.workflow
          ? [...APPLICATION_FIELDS, leaveCan.value.workflow.workflow_state_field]
          : APPLICATION_FIELDS
      ),
      order_by: 'from_date desc',
      limit: String(leaveCan.value.page_length || 20),
    }),
    immediate: false,
  })
  watch(
    () => myEmployee.value?.name,
    (name) => {
      if (name) applications.reload()
    },
    { immediate: true }
  )
  return applications
}

/**
 * Applications this user has to decide, or ones already decided.
 *
 * What counts as either is the server's to say — see `get_leave_approval_queue`.
 * Building the filters here instead let "waiting on you" mean one thing on this
 * page and a slightly different thing in the badge counting it.
 */
export function useLeaveApprovalQueue(decided: MaybeRefOrGetter<boolean>) {
  const queue = useCall<LeaveApplicationRow[], { decided: number }>({
    url: '/api/v2/method/tbs_commons.leave.api.get_leave_approval_queue',
    params: () => ({ decided: toValue(decided) ? 1 : 0 }),
    immediate: false,
  })
  watch(
    () => toValue(decided),
    () => queue.reload(),
    { immediate: true }
  )
  return queue
}

export interface LeaveAllocationSummary {
  total_leaves: number
  expired_leaves: number
  leaves_taken: number
  leaves_pending_approval: number
  remaining_leaves: number
}

export interface LeaveDetails {
  leave_allocation: Record<string, LeaveAllocationSummary>
  leave_approver: string | null
  lwps: string[]
}

/**
 * Balances per leave type, from HRMS's own calculation — it accounts for
 * carry-forward, expiry and pending applications, none of which is derivable
 * from the allocation total alone.
 */
export function useLeaveDetails(
  employee: MaybeRefOrGetter<string | undefined>
) {
  return useCall<LeaveDetails, { employee: string; date: string }>({
    url: '/api/v2/method/hrms.hr.doctype.leave_application.leave_application.get_leave_details',
    params: () => ({
      employee: toValue(employee) ?? '',
      date: new Date().toISOString().slice(0, 10),
    }),
    immediate: false,
  })
}

/** Working days a range costs, per HRMS: holidays and half days included. */
export function useLeaveDayCount() {
  return useCall<
    number,
    {
      employee: string
      leave_type: string
      from_date: string
      to_date: string
      half_day: 0 | 1
      half_day_date?: string
    }
  >({
    url: '/api/v2/method/hrms.hr.doctype.leave_application.leave_application.get_number_of_leave_days',
    immediate: false,
  })
}

/**
 * Settle an application, by whatever route the site has configured — a Workflow
 * transition where one is running, a status change and a submit where none is.
 * One call either way, because the two halves are one action for the approver.
 */
export function useLeaveDecision() {
  return useCall<
    { name: string; status: string; docstatus: number },
    { name: string; decision: string }
  >({
    url: '/api/v2/method/tbs_commons.leave.api.decide_leave_application',
    method: 'POST',
    immediate: false,
  })
}

/**
 * Raise a leave application for the employee behind this session.
 *
 * Through a whitelisted method rather than the document API: which fields a
 * request may set, and which employee it is for, are the server's to decide —
 * see `request_leave`. A form that posts a document decides both for itself.
 */
export function useRequestLeave() {
  return useCall<{ name: string }, { doc: string }>({
    url: '/api/v2/method/tbs_commons.leave.api.request_leave',
    method: 'POST',
    immediate: false,
  })
}

// No workflow description is fetched here, deliberately. Every row already
// arrives carrying its own label, style and permitted actions — the server
// settles all three, because `status` alone is not the answer and a site running
// a Workflow styles its own states — so nothing on these pages ever read the
// workflow itself. Reloading it after a decision refreshed nothing; reloading
// the queue is what moves the labels.

export interface LeaveStatusDisplay {
  label: string
  theme: BadgeTheme
}

/**
 * How a leave application reads to a person.
 *
 * An approvals-queue row arrives already settled -- the server decorates those,
 * because what a reviewer may do to a row is a permission question and has to be
 * answered there anyway -- and this just maps the style to a badge.
 *
 * A row the page fetched for itself arrives with neither, and is labelled here.
 * Worth being explicit about what that does and does not move off the server:
 * every *name* and every *style* is still the site's, read out of the workflow
 * states and the decision vocabulary the permissions endpoint sends. What is
 * stated here is the docstatus rule -- 2 is cancelled, 0 is still pending
 * whatever `status` says -- which is Frappe's document lifecycle rather than
 * anything a site configures.
 */
export function leaveStatus(
  row: LeaveApplicationRow,
  permissions: LeavePermissions = leaveCan.value
): LeaveStatusDisplay {
  if (row.status_label) {
    return { label: row.status_label, theme: styleTheme(row.status_style) }
  }

  const workflow = permissions.workflow
  if (workflow) {
    const state = row[workflow.workflow_state_field] as string | undefined
    const style =
      workflow.states.find((row) => row.state === state)?.style ?? null
    return { label: state || row.status, theme: styleTheme(style) }
  }

  if (row.docstatus === 2) return { label: 'Cancelled', theme: styleTheme(null) }
  if (row.docstatus === 0) return { label: 'Pending', theme: styleTheme('Warning') }

  // The outcome's own styling, as the server sent it -- so a site that renamed
  // an outcome or restyled it gets that here without an edit.
  const decision = permissions.decisions.find((row2) => row2.value === row.status)
  return { label: row.status, theme: styleTheme(decision?.style ?? null) }
}

/** Wording, and only wording. An outcome this does not know keeps the server's
 *  name for it, which is what a site that added one would want it called. */
const DECISION_LABELS: Record<string, string> = {
  Approved: 'Approve',
  Rejected: 'Deny',
}

/** How a leave decision button reads. The shape and the one-solid-button rule
 *  are shared (`workflowStyle`); leave supplies only its own wording. */
export function decisionButton(decision: LeaveDecision): DecisionButton {
  return buildDecisionButton(decision, DECISION_LABELS)
}

/** The buttons for one row: the outcomes the server said this row accepts. */
export function decisionButtons(
  offered: string[] | undefined,
  vocabulary: LeaveDecision[]
): DecisionButton[] {
  return buildDecisionButtons(offered, vocabulary, DECISION_LABELS)
}

File: commons/requests/leave.py
"""Leave: what is its own, and the endpoints the frontend reaches.

Leave is self-service: an employee requests their own, and an approver decides
it. `request_leave` raises one and `decide_leave_application` settles one;
HRMS's own validation is what a request has to get past either way.

The shape of all that -- who may decide, which rows are waiting on you, how
many there are, what the outcomes are called and how each one reads -- is
`approvals.RequestType`, shared with expenses and read by procurement. What is
left here is what is genuinely leave's:

* the doctype, its deciding field and the two values on it that are not a
  decision;
* a backstop derived from the doctype's own permission rows, which leave can do
  and expenses cannot -- `Leave.admin_roles`;
* HRMS refusing an approver's own *approval* rather than their own submit, so
  preventing self-approval costs one outcome and not all of them --
  `self_approval_outcome`.

Decisions come from the active Frappe Workflow when the site runs one. HRMS
expects that -- `validate_for_self_approval` steps aside for a workflow -- and
procurement has worked that way from the start, so a site that configures one
here gets its states, its styling and its transitions honoured rather than
bypassed. Without a workflow the outcomes are read off the `status` field's own
options, so a site that adds one by Property Setter is offered it. Neither path
has a state name or a role name behind it in this file.
"""

import frappe

from commons.commons_core.doc_perms import roles_with_permission
from commons.requests import approvals

LEAVE_APPLICATION = "Leave Application"


class Leave(approvals.RequestType):
	doctype = LEAVE_APPLICATION
	approver_field = "leave_approver"
	approver_name_field = "leave_approver_name"

	# Exactly the two `LeaveApplication.on_submit` refuses to submit: `Open` is
	# where an application starts and `Cancelled` is where cancelling it lands.
	non_decision_values = ("Open", "Cancelled")

	# HRMS will not let an employee apply *this* outcome to their own
	# application -- see `LeaveApplication.validate_for_self_approval`, which
	# tests `self.status == "Approved"` and so blocks approving but not turning
	# down. Expenses have no equivalent; see `approvals.self_approval_outcome`.
	self_approval_setting = "prevent_self_leave_approval"
	self_approval_outcome = "Approved"

	approver_mandatory_setting = "leave_approver_mandatory_in_leave_application"

	# Soonest first while they are still decisions to make; most recently
	# touched first once they are history.
	pending_order_by = "from_date asc"
	decided_order_by = "modified desc"

	request_fields = (
		"leave_type",
		"from_date",
		"to_date",
		"half_day",
		"half_day_date",
		"description",
		"leave_approver",
	)

	list_fields = (
		"name",
		"employee",
		"employee_name",
		"leave_type",
		"from_date",
		"to_date",
		"total_leave_days",
		"half_day",
		"description",
		"status",
		"docstatus",
		"leave_approver",
		"leave_approver_name",
		"posting_date",
	)

	def admin_roles(self) -> set[str]:
		"""Roles that own the doctype, and so back up an absent approver.

		Derived from the doctype's own permission rows rather than named here,
		so a site that renames its HR roles or grants the backstop to another one
		says so once, in the Role Permission Manager, and this follows.

		The test is submit *and* create at once. Deciding is what `submit`
		grants, and on the standard rows `Leave Approver` holds it too -- for
		every application on the site, which is exactly the over-reach this
		narrows. A role that may also raise applications is one that owns the
		doctype rather than one the workflow points at, and that is the backstop
		wanted here.

		Empty is a legitimate answer: a site whose permissions name no such role
		has no backstop, and the named approver is then the only person who can
		decide.
		"""
		return roles_with_permission(self.doctype, submit=1, create=1)

	def row_is_theirs(self, row, admin: bool) -> bool:
		"""This application's own approver, or one of the roles that own it.

		Narrower than the doctype right on purpose: `has_permission("submit")`
		answers "may this user submit leave applications", which for the
		`Leave Approver` role means all of them.
		"""
		return admin or row.get(self.approver_field) == frappe.session.user

	def queue_predicate(self, decided: bool, admin: bool) -> tuple[dict, list | None]:
		"""Which applications are this user's to look at, deciding or decided.

		An undecided one is `Open` as well as unsubmitted, which is the state
		HRMS leaves a draft in -- without it a status somebody set from the desk
		without submitting would show up in the queue with buttons on it and be
		missing from the badge beside it.

		An admin sees everyone's, which is what makes them a usable backstop;
		anyone else sees the ones that name them.
		"""
		filters = {"docstatus": 1 if decided else 0}
		if not decided:
			filters["status"] = "Open"
		if not admin:
			filters[self.approver_field] = frappe.session.user
		return filters, None

	def no_employee_message(self) -> str:
		return frappe._(
			"Your login is not linked to an employee record, so leave cannot be requested for it."
		)

	def foreign_employee_message(self) -> str:
		return frappe._("Leave can only be requested for your own employee record here.")

	def permlevel_message(self) -> str:
		return frappe._(
			"You are not permitted to decide leave applications. The Leave Approver role grants this."
		)

	def not_yours_message(self, doc) -> str:
		return frappe._("{0} is not your leave application to decide — it is assigned to {1}.").format(
			doc.name,
			doc.get(self.approver_name_field) or doc.get(self.approver_field) or frappe._("nobody"),
		)

	def already_decided_message(self, doc) -> str:
		return frappe._("{0} has already been {1}.").format(doc.name, doc.status.lower())


LEAVE = Leave()


@frappe.whitelist()
def get_leave_permissions() -> dict:
	"""What the session user may do with leave, plus their approval backlog.

	See `approvals.RequestType.permissions`, which is this answer and the
	argument for every field in it.
	"""
	return LEAVE.permissions()


@frappe.whitelist()
def get_leave_approval_queue(decided: int = 0) -> list[dict]:
	"""Applications waiting on this user's decision, or ones already decided.

	An admin looking at somebody else's queue gets buttons, a reader who merely
	has the page open does not, and an approver looking at their own application
	does not get the one HRMS would refuse. See
	`approvals.RequestType.approval_queue`.
	"""
	return LEAVE.approval_queue(bool(frappe.utils.cint(decided)))


@frappe.whitelist(methods=["POST"])
def request_leave(doc: str | dict) -> dict:
	"""Raise a leave application for the employee behind this session.

	Through here rather than straight at the document API so the field list a
	request may set is the server's, and so the employee is resolved from the
	session instead of accepted -- see `approvals.RequestType.request_employee`.
	"""
	values = LEAVE.parse_request(doc)
	employee = LEAVE.request_employee(values, ["name"])
	return frappe.get_doc(LEAVE.new_request(values, employee)).insert().as_dict()


@frappe.whitelist(methods=["POST"])
def decide_leave_application(name: str, decision: str) -> dict:
	"""Settle a leave application, by the route the site has configured.

	See `approvals.RequestType.decide`: a Workflow transition where one is
	running, and a status change plus a submit in one transaction where none is.
	"""
	return LEAVE.decide(name, decision)
